README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# Overview <a href='https://lmjl-alea.github.io/flipr/'><img src='man/figures/logo.png' align="right" height="139" /></a>

<!-- badges: start -->

[![R-CMD-check](https://github.com/LMJL-Alea/flipr/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/LMJL-Alea/flipr/actions/workflows/R-CMD-check.yaml)
[![test-coverage](https://github.com/LMJL-Alea/flipr/workflows/test-coverage/badge.svg)](https://github.com/LMJL-Alea/flipr/actions)
[![Codecov test
coverage](https://codecov.io/gh/LMJL-Alea/flipr/branch/master/graph/badge.svg)](https://codecov.io/gh/LMJL-Alea/flipr?branch=master)
[![pkgdown](https://github.com/LMJL-Alea/flipr/workflows/pkgdown/badge.svg)](https://github.com/LMJL-Alea/flipr/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/flipr)](https://CRAN.R-project.org/package=flipr)
<!-- badges: end -->

The goal of the [**flipr**](https://lmjl-alea.github.io/flipr/) package
is to provide a flexible framework for making inference via permutation.
The idea is to promote the permutation framework as an incredibly
well-suited tool for inference on complex data. You supply your data, as
complex as it might be, in the form of lists in which each entry stores
one data point in a representation that suits you and
[**flipr**](https://lmjl-alea.github.io/flipr/) takes care of the
permutation magic and provides you with either point estimates or
confidence regions or $p$-value of hypothesis tests. Permutation tests
are especially appealing because they are exact no matter how small or
big your sample sizes are. You can also use the so-called
*non-parametric combination* approach in this setting to combine several
statistics to better target the alternative hypothesis you are testing
against. Asymptotic consistency is also guaranteed under mild conditions
on the statistic you use. The
[**flipr**](https://lmjl-alea.github.io/flipr/) package provides a
flexible permutation framework for making inference such as point
estimation, confidence intervals or hypothesis testing, on any kind of
data, be it univariate, multivariate, or more complex such as
network-valued data, topological data, functional data or density-valued
data.

## Installation

You can install the latest stable version of
[**flipr**](https://lmjl-alea.github.io/flipr/) on CRAN with:

``` r
install.packages("flipr")
```

Or you can install the development version from
[GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("LMJL-Alea/flipr")
```

## Example

``` r
library(flipr)
```

We hereby use the very simple t-test for comparing the means of two
univariate samples to show how easy it is to carry out a permutation
test with [**flipr**](https://lmjl-alea.github.io/flipr/).

### Data generation

Let us first generate two samples of size $15$ governed by Gaussian
distributions with equal variance but different means:

``` r
set.seed(123)
n <- 15
x <- rnorm(n = n, mean = 0, sd = 1)
y <- rnorm(n = n, mean = 1, sd = 1)
```

Given the data we simulated, the parameter of interest here is the
difference between the means of the distributions, say
$\delta = \mu_y - \mu_x$.

### Make the two samples exchangeable under $H_0$

In the context of null hypothesis testing, we consider the null
hypothesis $H_0: \mu_y - \mu_x = \delta$. We can use a permutation
scheme to approach the $p$-value if the two samples are *exchangeable*
under $H_0$. This means that we need to transform for example the second
sample to *make* it exchangeable with the first sample under $H_0$. In
this simple example, this can be achieved as follows. Let
$X_1, \dots, X_{n_x} \sim \mathcal{N}(\mu_x, 1)$ and
$Y_1, \dots, Y_{n_y} \sim \mathcal{N}(\mu_y, 1)$. We can then transform
the second sample as $Y_i \longleftarrow Y_i - \delta$.

We can define a proper function to do this, termed the *null
specification* function, which takes two input arguments:

- `y` which is a list storing the data points in the second sample;
- `parameters` which is a numeric vector of values for the parameters
  under investigation (here only $\delta$ and thus `parameters` is of
  length $1$ with `parameters[1] = delta`).

In our simple example, it boils down to:

``` r
null_spec <- function(y, parameters) {
  purrr::map(y, ~ .x - parameters[1])
}
```

### Choose suitable test statistics

Next, we need to decide which test statistic(s) we are going to use for
performing the test. Here, we are only interested in one parameter,
namely the mean difference $\delta$. Since the two samples share the
same variance, we can use for example the $t$-statistic with a pooled
estimate of the common variance.

This statistic can be easily computed using
`stats::t.test(x, y, var.equal = TRUE)$statistic`. However, we want to
extend its evaluation to any permuted version of the data. Test
statistic functions compatible with
[**flipr**](https://lmjl-alea.github.io/flipr/) should have at least two
mandatory input arguments:

- `data` which is either a concatenated list of size $n_x + n_y$
  regrouping the data points of both samples or a distance matrix of
  size $(n_x + n_y) \times (n_x + n_y)$ stored as an object of class
  `dist`.
- `indices1` which is an integer vector of size $n_x$ storing the
  indices of the data points belonging to the first sample in the
  current permuted version of the data.

Some test statistics are already implemented in
[**flipr**](https://lmjl-alea.github.io/flipr/) and ready to use.
User-defined test statistics can be used as well, with the use of the
helper function `use_stat(nsamples = 2, stat_name = )`. This function
creates and saves an `.R` file in the `R/` folder of the current working
directory and populates it with the following template:

``` r
#' Test Statistic for the Two-Sample Problem
#'
#' This function computes the test statistic...
#'
#' @param data A list storing the concatenation of the two samples from which
#'   the user wants to make inference. Alternatively, a distance matrix stored
#'   in an object of class \code{\link[stats]{dist}} of pairwise distances
#'   between data points.
#' @param indices1 An integer vector that contains the indices of the data
#'   points belong to the first sample in the current permuted version of the
#'   data.
#'
#' @return A numeric value evaluating the desired test statistic.
#' @export
#'
#' @examples
#' # TO BE DONE BY THE DEVELOPER OF THE PACKAGE
stat_{{{name}}} <- function(data, indices1) {
  n <- if (inherits(data, "dist"))
    attr(data, "Size")
  else if (inherits(data, "list"))
    length(data)
  else
    stop("The `data` input should be of class either list or dist.")

  indices2 <- seq_len(n)[-indices1]

  x <- data[indices1]
  y <- data[indices2]

  # Here comes the code that computes the desired test
  # statistic from input samples stored in lists x and y

}
```

For instance, a
[**flipr**](https://lmjl-alea.github.io/flipr/)-compatible version of
the $t$-statistic with pooled variance will look like:

``` r
my_t_stat <- function(data, indices1) {
  n <- if (inherits(data, "dist"))
    attr(data, "Size")
  else if (inherits(data, "list"))
    length(data)
  else
    stop("The `data` input should be of class either list or dist.")

  indices2 <- seq_len(n)[-indices1]

  x <- data[indices1]
  y <- data[indices2]
  
  # Here comes the code that computes the desired test
  # statistic from input samples stored in lists x and y
  x <- unlist(x)
  y <- unlist(y)
  
  stats::t.test(x, y, var.equal = TRUE)$statistic
}
```

Here, we are only going to use the $t$-statistic for this example, but
we might be willing to use more than one statistic for a parameter or we
might have several parameters under investigation, each one of them
requiring a different test statistic. We therefore group all the test
statistics that we need into a single list:

``` r
stat_functions <- list(my_t_stat)
```

### Assign test statistics to parameters

Finally we need to define a named list that tells
[**flipr**](https://lmjl-alea.github.io/flipr/) which test statistics
among the ones declared in the `stat_functions` list should be used for
each parameter under investigation. This is used to determine bounds on
each parameter for the plausibility function. This list, often termed
`stat_assignments`, should therefore have as many elements as there are
parameters under investigation. Each element should be named after a
parameter under investigation and should list the indices corresponding
to the test statistics that should be used for that parameter in
`stat_functions`. In our example, it boils down to:

``` r
stat_assignments <- list(delta = 1)
```

### Use the plausibility function

Now we can instantiate a plausibility function as follows:

``` r
pf <- PlausibilityFunction$new(
  null_spec = null_spec,
  stat_functions = stat_functions,
  stat_assignments = stat_assignments,
  x, y
)
#> ! Setting the seed for sampling permutations is mandatory for obtaining a continuous p-value function. Using `seed = 1234`.
```

Now, assume we want to test the following hypotheses:

$$ H_0: \delta = 0 \quad \mbox{v.s.} \quad H_1: \delta \ne 0. $$

We use the `$get_value()` method for this purpose, which essentially
evaluates the permutation $p$-value of a two-sided test by default:

``` r
pf$get_value(0)
#> [1] 0.1078921
```

We can compare the resulting $p$-value with the one obtained using the
more classic parametric test:

``` r
t.test(x, y, var.equal = TRUE)$p.value
#> [1] 0.1030946
```

The permutation $p$-value does not quite match the parametric one. This
is because of two reasons:

1.  The resolution of a permutation $p$-value is of the order of
    $1/(B+1)$, where $B$ is the number of sampled permutations. By
    default, the plausibility function is instantiated with $B = 1000$:

``` r
pf$nperms
#> [1] 1000
```

2.  We randomly sample $B$ permutations out of the
    $\binom{n_x+n_y}{n_x}$ possible permutations and therefore introduce
    extra variability in the $p$-value.

If we were to ask for more permutations, say $B = 1,000,000$, we would
be much closer to the parametric $p$-value:

``` r
pf$set_nperms(1000000)
pf$get_value(0)
#> [1] 0.1029879
```


NEWS.md
# flipr (development version)

# flipr 0.3.3

- Updated URLs in `DESCRIPTION` file;
- Properly documented package overview in `flipr-package.R` file;
- Updated GitHub Action workflows;
- Removed suggested dependency to the entire tidyverse;
- Updated maintainer email.

# flipr 0.3.2

- Removed `akima` from package dependencies.
- Updated GHA workflows and `roxygen2` version.
- Updated all URLs to reflect change of ownership.

# flipr 0.3.1

- Remove any occurrence of base-R pipe operator in favor of [**magrittr**](https://magrittr.tidyverse.org) pipe operator for now for backward compatibility with older versions of R.
- Improved way of handling statistics that rely on inter-point distances.
- Method `$get_value()` gains optional arguments `keep_null_distribution` and `keep_permutations` in case the user would like to keep track of either the permutation null distribution or the sampled permutations that have been used to produce it.
- Method `$get_value()` now handles extra parameters to be passed to the test statistics.

# flipr 0.3.0

## Breaking changes

The API has been changed:

* An [R6](https://r6.r-lib.org) class is now used to define the plausibility function, with methods that implement point estimation, confidence intervals and evaluation on grid for later plotting.
* The plotting capability is still outside the class but might move into the class methods as well in the future.

## Minor changes

* Test statistics based on Inter-point distances have been added and implemented in C++ through the [**Rcpp**](http://dirk.eddelbuettel.com/code/rcpp.html) package.
* Test coverage is set and achieves 20% coverage for now (#3, @C-Juliette).
* @C-Juliette has been added as contributor to the package.
* The p-value function has been consistently renamed plausibility function to better reflects what it stands for and avoid confusion with other definitions of the p-value function in the area of functional data analysis.

# flipr 0.2.1

* Fix warning when installing the package on Oracle Solaris 10, x86, 32 bit,
R-release due to pandoc version not available preventing the `pvalue-function` vignette from knitting properly.

# flipr 0.2.0

## New features

* Compute p-value surfaces when inferring multiple parameters at once via
`two_sample_pf()`;
* Compute a point estimation for a single parameter of interest via
`two_sample_pe()`;
* Compute a confidence interval for a single parameter of interest via
`two_sample_ci()`;
* Automatically draw the p-value function for a single parameter of interest via
`two_sample_viz()`.

## Full list of changes

* Change argument names `test` and `combining_function` to `type` and
`combine_with` respectively. Add argument `alternative` to specify which kind of
test is performed.
* Add a function to compute p-value functions for sets of null hypotheses.
* Draft of article illustrating the computation of p-value functions with
[**flipr**](https://lmjl-alea.github.io/flipr/).
* Add $t$, mean and Fisher test statistics.
* Correct two-tail p-value computation.
* Better API for pvalue function.
* Added a function for two sample confidence interval for a single parameter.
* Added an article about permutation inference overview.
* Added dependency to the [**ggplot2**](https://ggplot2.tidyverse.org) package
for plotting features.
* Added dependency to the [**cli**](https://cli.r-lib.org) package for
displaying important information to the user in an elegant fashion.
* Added function for point estimation of a single parameter.
* Added function for confidence interval of a single parameter.
* Added function for automatic plotting of the p-value function of a single
parameter.
* Added dependency to the [**withr**](https://withr.r-lib.org) package for
properly handle local random seed generator settings.
* Added dependency to the
[**viridisLite**](https://github.com/sjmgarnier/viridisLite) package for using
colors compatible with most common types of color blindness.

# flipr 0.1.1

* Fix warning when installing the package on Oracle Solaris 10, x86, 32 bit,
R-release due to pandoc version not available.
* Switch to GPL-3 license because **purrr** dependency is GPL-3 licensed.

# flipr 0.1.0

* Initial release.
* Added a `NEWS.md` file to track changes to the package.


R/two-sample-pf.R
#' Two-Sample Permutation P-Value Function
#'
#' This function calculates the permutation p-value function for two-sample
#' problems. This is done through the specification of a set of null hypotheses
#' of the form `F_X = F_{g(Y, parameters)}` where `g` is a user-supplied
#' function.
#'
#' @param parameters A list of vectors specifying a set of candidate parameter
#'   values under the null hypothesis.
#' @param null_specification A function with two arguments `y` and `parameters`
#'   such that `F_X = F_{null_specification(Y, parameters)}` under the null
#'   hypothesis.
#' @inheritParams two_sample_test
#'
#' @return A vector of p-values of the permutation test for each parameter
#'   values specified through the first argument.
#' @export
#'
#' @examples
#' x1 <- rnorm(10)
#' x2 <- rnorm(10, 3)
#' null_spec <- function(y, parameters) {y - parameters[1]}
#' two_sample_pf(
#'   parameters = 3,
#'   null_specification = null_spec,
#'   x = x1,
#'   y = x2,
#'   statistic = stat_t,
#'   seed = 1234,
#'   B = 1000,
#'   alternative = "two_tail"
#' )
two_sample_pf <- function(parameters,
                          null_specification,
                          x, y,
                          statistic = stat_hotelling,
                          B = 1000L,
                          alternative = "right_tail",
                          combine_with = "tippett",
                          type = "exact",
                          seed = NULL) {
  if (!is.null(seed)) withr::local_seed(seed)
  null_spec <- rlang::as_function(null_specification)
  parameters %>%
    purrr::map_dbl(~ {
      y <- null_spec(y, .x)
      two_sample_test(
        x = x,
        y = y,
        statistic = statistic,
        B = B,
        type = type,
        alternative = alternative,
        combine_with = combine_with
      )$pvalue
    })
}

two_sample_pf_new <- function(null_specification,
                              x, y,
                              max_confidence_level = NULL,
                              statistic = stat_hotelling,
                              B = 1000L,
                              alternative = "right_tail",
                              combine_with = "tippett",
                              type = "exact",
                              seed = NULL) {
  n_params <- length(statistic)
  null_specification <- rlang::as_function(null_specification)

  if (is.null(seed)) {
    seed <- 1234
    cli::cli_alert_info(
      "Setting the seed for sampling permutations is mandatory for obtaining a continuous p-value function. Using `seed = 1234`."
    )
  }

  pvf_exact <- setup_pvf(
    null_spec = null_specification,
    x = x,
    y = y,
    statistic = statistic,
    B = B,
    type = type,
    alternative = alternative,
    combine_with = combine_with,
    seed = seed
  )

  if (is.null(max_confidence_level)) return(pvf_exact)

  writeLines(paste("- Computing point estimates for all", n_params, "parameters..."))
  genoud_results <- rgenoud::genoud(
    fn = pvf_exact,
    nvars = n_params,
    max = TRUE,
    pop.size = 50,
    max.generations = 20,
    wait.generations = 5,
    BFGSburnin = 5
  )

  Bmin <- 10 * (n_params - 1 + max_confidence_level) / (1 - max_confidence_level)
  if (B < Bmin) {
    cli::cli_alert_info(paste0(
      "The required number of permutations (B = ",
      B,
      ") is insufficient for reaching the input maximum confidence level (",
      round(max_confidence_level * 100, digits = 2),
      "%). Setting it to B = ",
      Bmin,
      "."
    ))
    B <- as.integer(Bmin)
  }

  writeLines("- Setting up individual null specifications...")
  null_specs <- 1:n_params %>%
    purrr::map(~ function(y, parameters) {
      all_parameters <- genoud_results$par
      all_parameters[.x] <- parameters[1]
      null_specification(y, all_parameters)
    })

  writeLines("- Computing evaluation grid...")
  grid_list <- 1:n_params %>%
    purrr::map(~ {
      null_spec <- null_specs[[.x]]
      stat <- statistic[[.x]]
      pvf_temp <- setup_pvf(
        null_spec = null_spec,
        x = x,
        y = y,
        statistic = stat,
        B = B,
        type = type,
        alternative = "two_tail",
        seed = seed
      )

      writeLines(paste(
        "- Computing a point estimate for parameter",
        .x,
        "..."
      ))
      pe <- two_sample_pe_new(pvf_temp)

      writeLines(paste(
        "- Computing a confidence interval with confidence level",
        max_confidence_level,
        "for parameter",
        .x,
        "..."
      ))
      ci <- two_sample_ci_new(
        pvf_temp,
        alpha = (1 - max_confidence_level) / n_params,
        point_estimate = pe
      )

      c(pe, ci)
    })

  grid_size <- 20
  grid_out <- (grid_size + 1)^n_params

  if (grid_out > 1000)
    abort(
      "The output grid size is too large and will take too long to compute. Please consider either a coarser grid or inferring less parameters or reducing the number of permutations (it is though recommended that B > 1000)."
    )

  writeLines(paste(
    "- Computing the exact p-value function on a grid of size",
    grid_out,
    "..."
  ))
  x_in <- generate_grid(grid_list, n = grid_size)
  y_in <- x_in %>%
    purrr::array_tree(margin = 1) %>%
    purrr::map_dbl(pvf_exact)

  writeLines("- Computing Kriging model for interpolation using the DiceKriging::km() function...")
  km <- DiceKriging::km(
    design = x_in,
    response = qnorm(y_in),
    nugget = sqrt(.Machine$double.eps),
    control = list(trace = FALSE),
    covtype = "exp"
  )

  pvf_approx <- function(parameters) {
    interp <- predict(
      km,
      newdata = rbind(parameters),
      type = "UK",
      se.compute = FALSE,
      checkNames = FALSE
    )
    pnorm(interp$mean)
  }

  setup_pvf(
    null_spec = null_specification,
    x = x,
    y = y,
    statistic = statistic,
    B = B,
    type = type,
    alternative = alternative,
    combine_with = combine_with,
    seed = seed,
    pvf = pvf_approx,
    max_confidence_level = max_confidence_level,
    design = x_in,
    response = y_in
  )
}

setup_pvf <- function(null_spec, x, y,
                      statistic = stat_t,
                      B = 1000,
                      type = "exact",
                      alternative = "two_tail",
                      combine_with = "tippett",
                      seed = NULL,
                      pvf = NULL,
                      max_confidence_level = NULL,
                      design = NULL,
                      response = NULL) {
  if (is.null(pvf)) {
    pvf <- function(parameters) {
      withr::local_seed(seed)
      y <- null_spec(y, parameters)
      two_sample_test(
        x = x,
        y = y,
        statistic = statistic,
        B = B,
        type = type,
        alternative = alternative,
        combine_with = combine_with
      )$pvalue
    }
  }

  attr(pvf, "null_spec") <- null_spec
  attr(pvf, "x") <- x
  attr(pvf, "y") <- y
  attr(pvf, "statistic") <- statistic
  attr(pvf, "B") <- B
  attr(pvf, "type") <- type
  attr(pvf, "alternative") <- alternative
  attr(pvf, "combine_with") <- combine_with
  attr(pvf, "seed") <- seed
  attr(pvf, "max_confidence_level") <- max_confidence_level
  attr(pvf, "design") <- design
  attr(pvf, "response") <- response

  class(pvf) <- "pvalue_function"
  pvf
}
